File: merkletree.py
import hashlib
import random
import math

#isPoweroftwo is from geeks to geeks
# Function to check
# Log base 2
def Log2(x):
    if x == 0:
        return false;

    return (math.log10(x) /
            math.log10(2));

# Function to check
# if x is power of 2
def isPowerOfTwo(n):
    return (math.ceil(Log2(n)) ==
            math.floor(Log2(n)));

class Node:
    def __init__(self, left, right, is_leaf=False):
        self.left = left
        self.right = right
        self.parent = None
        self.is_leaf = is_leaf
        self.entry = None

    def teachKids(self, left, right):
        self.left.parent = self
        self.right.parent = self

    def makeEntry(self, key, value):
        entry = Entry(value, key)
        self.entry = entry

    def printNodePointers(self):
        node = self.entry.value
        if self.parent == None:
            p = "no pointer"
            rs = "none"
            ls = "none"
            gp = 'none'
        else:
            if self.parent.right == None:
                rs = 'none'
            else:
                rs = self.parent.right.entry.value
            if self.parent.left == None:
                ls = 'none'
            else:
                ls = self.parent.left.entry.value
            if self.parent.parent == None:
                gp = 'none'
            else:
                gp = self.parent.parent.entry.value
            p = self.parent.entry.value
        if self.right == None:
            r = "no pointer"
        else:
            r = self.right.entry.value
        if self.left == None:
            l = "no pointer"
        else:
            l = self.left.entry.value
        sibs = (". Right sib: " + rs + ". Left Sib: " + ls + ". grandparent: " + gp)
        essential = ("This is node: " + node + ". Parent: " + p + ". Right: " + r + ". Left: " + l)
        print(essential + sibs)

    def teachParentMyName(self, is_right):
        if self.parent:
            if is_right:
                self.parent.right = self
            else:
                self.parent.left = self

class Entry:
    def __init__(self, value, key=None):
        self.value = value
        self.key = key

    def makeKey(self):
        #makes key from value, should only be used when making a leaf
        hash = hashlib.sha512(self.value.encode()).hexdigest()
        self.key = hash

    def toString(self):
        return "{key: "+self.key+", value: "+self.value+"}"

class MerkleTree:
    def __init__(self):
        self.RootNode = None #node class
        self.RootHash = None #hex decimal hash
        self.entries = [] #list of entries
        self.entries_map = {} #map entry to node
        self.node_map = {} #map hash to node

    def makeLeafNode(self, entry):
        #makes a leaf node and updates the entries
        #entry.makeKey()
        self.entries.append(entry)
        #hash = entry.key#hashlib.sha512(entry.value.encode()).hexdigest()
        new_node = Node( None, None, True)
        new_node.entry = entry
        self.entries_map[entry] = new_node
        self.node_map[entry.key] = new_node
        return new_node

    def updateRoot(self, node):
        self.RootNode = node
        self.RootHash = node.entry.key
        return self.RootHash

    def checkIfGranparentRoot(self, node, new_nd):
        #check if grandparent is root, if not asigns proper parent to relevant node
        if node.parent.parent != None: #parent is not root
            new_nd.parent = node.parent.parent
        else:
            self.updateRoot(new_nd)

    def Insert(self, entry):

        new_leaf_node = self.makeLeafNode(entry)
        #check if node is about to have only 1 leaf
        if len(self.entries) == 1: # first insert
            return self.updateRoot(new_leaf_node)

        #tree is perfect before insert or is about to have 2 leaves
        elif isPowerOfTwo(len(self.entries)-1) or len(self.entries) == 2:
            future_sib = self.RootNode
            str = future_sib.entry.key + new_leaf_node.entry.key
#            str = future_sib.entry.toString() + new_leaf_node.entry.toString()
            new_root_hash = hashlib.sha512(str.encode()).hexdigest()
            new_val = future_sib.entry.value + new_leaf_node.entry.value
            new_root = Node(future_sib, new_leaf_node)
            new_root.makeEntry(new_root_hash, new_val)

            new_root.teachKids(future_sib, new_leaf_node)
            self.node_map[new_root_hash] = new_root

            return self.updateRoot(new_root)

        second_to_last_entry = self.entries[-2]
        temp_root = self.entries_map[second_to_last_entry]

        #initial insert and there is no lonely leaf but not balance
        if temp_root.parent.left.is_leaf and temp_root.is_leaf:
            str = temp_root.parent.entry.key + new_leaf_node.entry.key
#            str = temp_root.parent.entry.toString() + new_leaf_node.entry.toString()
            new_nd_h = hashlib.sha512(str.encode()).hexdigest()

            new_val = temp_root.parent.entry.value + new_leaf_node.entry.value
            new_nd = Node(temp_root.parent, new_leaf_node)
            new_nd.makeEntry(new_nd_h, new_val)

#            new_nd.makeEntry(new_nd_h, new_val)
            self.checkIfGranparentRoot(temp_root, new_nd)
            new_nd.teachKids(temp_root.parent, new_leaf_node)
            self.node_map[new_nd_h] = new_nd

            temp_root = new_nd

        #initial insert of leaf and there is already a lonely leaf
        elif temp_root.parent.left.is_leaf == False and temp_root.is_leaf:
            str = temp_root.entry.key + new_leaf_node.entry.key
#            str = temp_root.entry.toString() + new_leaf_node.entry.toString()
            new_nd_h = hashlib.sha512(str.encode()).hexdigest()

            new_val = temp_root.entry.value + new_leaf_node.entry.value
            new_nd = Node(temp_root, new_leaf_node)
            new_nd.makeEntry(new_nd_h, new_val)

            new_nd.parent = temp_root.parent
            new_nd.teachKids(temp_root, new_leaf_node)
            self.node_map[new_nd_h] = new_nd

            temp_root = new_nd

        #inserts after the initial insert
        while temp_root != self.RootNode:
            node_to_delete_key = temp_root.parent.entry.key
            str = temp_root.parent.left.entry.key + temp_root.entry.key
#            str = temp_root.parent.left.entry.toString() + new_leaf_node.entry.toString()
            new_nd_h = hashlib.sha512(str.encode()).hexdigest()

            new_val = temp_root.parent.left.entry.value + temp_root.entry.value
            new_nd = Node(temp_root.parent.left, temp_root)
            new_nd.makeEntry(new_nd_h, new_val)

            self.checkIfGranparentRoot(temp_root, new_nd)
            new_nd.teachKids(temp_root.parent.left, temp_root)
            self.node_map[new_nd_h] = new_nd
            del self.node_map[node_to_delete_key] #delete old parent

            temp_root = new_nd
        self.updateRoot(temp_root)
        return temp_root.entry.key

    def generateMerklePath(self, key):
        merkle_path = []
        if key in self.node_map:
            node = self.node_map[key]
        else:
            return "path_not_found"

        while node is not self.RootNode:
            if node.parent.right == node: #is right nodes
                merkle_path.append(node.parent.left.entry.key)
            else:
                merkle_path.append(node.parent.right.entry.key)
            node = node.parent
        return merkle_path

    def getRightMostNode(self):
        #returns right most node
        right_nd = self.RootNode
        while right_nd.is_leaf != True:
            right_nd = right_nd.right
        return right_nd

    def removeRightMostNode(self, right_nd):
        '''
        Removes right most node from its spot and returns it
        '''
        if right_nd.parent == self.RootNode:
            #special case 2 leafs
            self.RootNode = right_nd.parent.left
            self.RootNode.parent = None

        else:
            grandparent = right_nd.parent.parent
            grandparent.right = right_nd.parent.left
            right_nd.parent.left.parent = grandparent
        parent_to_delete = right_nd.parent.entry.key
        del self.node_map[parent_to_delete]
        return right_nd

    def deleteLeafFromMaps(self, node):
        del self.entries_map[node.entry]
        del self.node_map[node.entry.key]
        self.entries.remove(node.entry)

    def Delete(self, entry):
        '''
        The Delete function takes a key (Entry) as argument, traverses the
        Merkle Tree and finds that key. If the key exists, delete the
        corresponding Entry and re-balance the tree if necessary. Delete
        function will return updated root hash if the key was found otherwise
        "path_not_found” if the key doesn't exist.
        '''
        if entry not in self.entries_map:
            return 'path_not_found'

        delete_nd = self.entries_map[entry]
        right_nd = self.getRightMostNode()

        '''
        Below is logic block for special cases.
        Special case 1: only one node in tree prior to delete.
        Special case: 2: leafs in tree prior to delete.
        '''
        if len(self.entries_map) == 1:
            self.RootHash = None
            self.RootNode = None
            self.deleteLeafFromMaps(delete_nd)
            return self.RootNode
        elif len(self.entries_map) == 2:
            parent_to_delete = right_nd.parent.entry.key
            if right_nd == delete_nd:
                self.RootNode = right_nd.parent.left
            else:
                self.RootNode = right_nd
            self.RootNode.left, self.RootNode.right, self.RootNode.parent = None, None, None
            self.RootHash = self.RootNode.entry.key
            self.deleteLeafFromMaps(delete_nd)
            del self.node_map[parent_to_delete]
            return self.RootNode

        right_nd = self.removeRightMostNode(right_nd)
        temp_node = None
        if right_nd == delete_nd:
            #deleting the farthest right node
            temp_node = delete_nd.parent.left
        else:
            temp_node, temp_node.parent = right_nd, delete_nd.parent
            if delete_nd == delete_nd.parent.right:
                temp_node.parent.right = temp_node
            else:
                temp_node.parent.left = temp_node
        self.deleteLeafFromMaps(delete_nd)
        while temp_node != self.RootNode:
            is_right = True
            if temp_node.parent.parent:
                #check what side for teaching parent later
                if temp_node.parent == temp_node.parent.parent.right:
                    is_right = True
                if temp_node.parent == temp_node.parent.parent.left:
                    is_right = False

            node_to_delete_key = temp_node.parent.entry.key
            if temp_node == temp_node.parent.right:
                sibling = temp_node.parent.left
                str = sibling.entry.key + temp_node.entry.key
                new_val = sibling.entry.value + temp_node.entry.value
                new_nd_h = hashlib.sha512(str.encode()).hexdigest()

                new_nd = Node(temp_node.parent.left, temp_node)
                new_nd.makeEntry(new_nd_h, new_val)
                self.checkIfGranparentRoot(temp_node, new_nd)
                new_nd.teachKids(temp_node.parent.left, temp_node)
                #new_nd.teachParentMyName(is_right)

            elif temp_node == temp_node.parent.left:
                sibling = temp_node.parent.right
                str = temp_node.entry.key + sibling.entry.key
                new_val = temp_node.entry.value + sibling.entry.value
                new_nd_h = hashlib.sha512(str.encode()).hexdigest()

                new_nd = Node(temp_node, temp_node.parent.right)
                new_nd.makeEntry(new_nd_h, new_val)
                self.checkIfGranparentRoot(temp_node, new_nd)
                #above sets to new_nd to root if applicable
                new_nd.teachKids(temp_node, temp_node.parent.right)

            new_nd.teachParentMyName(is_right)
            self.node_map[new_nd_h] = new_nd
            del self.node_map[node_to_delete_key] #delete old parent
            temp_node = new_nd
            #if new_nd is root then loop stops
        return self.RootHash

    def VerifyMerklePath(self, entry, merkle_path):
        '''
        The VerifyMerklePath function takes a key (Entry) and its Merkle path,
        the ordered list of sibling hashes as argument. It computes all the
        hashes on the path from the given Entry to the root using the location
        and the MerklePath. The newly computed root hash is compared to the
        stored root for verification. Function returns true if the
        verification succeeds (if the newly computed root hash is equal to
        the stored root hash) otherwise return false.
        Arguments: Entry object, location (or index) of this Entry object in
        the Merkle Tree, and the list/array of hashes on Merkle Path
        Return: bool
        '''
        if entry not in self.entries_map.keys():
            return False
        node = self.entries_map[entry]
        parent_hash = node.entry.key
        for i in range(len(merkle_path)):
            if parent_hash not in self.node_map:
                #check if right or left
                return False
            if self.node_map[parent_hash] == self.node_map[parent_hash].parent.left:
                str = parent_hash + merkle_path[i]
            else:
                str = merkle_path[i] + parent_hash
            parent_hash = hashlib.sha512(str.encode()).hexdigest()
        return parent_hash == self.RootHash

File: merkletree.go
package main

import (
	"crypto/sha512"
	"encoding/hex"
	"fmt"
	"math"
)

// General utils
func getHash(str string) string {
	h := sha512.New()
	h.Write([]byte(str))
	out_hash := hex.EncodeToString(h.Sum(nil))
	return out_hash
}

func isPowerOfTwo(n int) bool {
	x := float64(n)
	if n == 0 {
		return false
	}
	return math.Ceil(math.Log2(x)) == math.Floor(math.Log2(x))
}

func removeEntry(l []*Entry, item *Entry) []*Entry {
	for i, other := range l {
		if other == item {
			return append(l[:i], l[i+1:]...)
		}
	}
	return nil
}

// Structs followed by there methods

type Node struct {
	left    *Node
	right   *Node
	is_leaf bool
	parent  *Node
	entry   *Entry
}

func NewNode(left *Node, right *Node, is_leaf bool) Node {
	new_nd := Node{left: left, right: right, is_leaf: is_leaf}
	return new_nd
}

func makeNilNode() *Node {
	nil_node := &Node{left: nil, right: nil, parent: nil}
	nil_node.makeEntry("", "")
	return nil_node
}

func (n *Node) makeEntry(key string, value string) {
	entry := &Entry{key: key, value: value}
	n.entry = entry
}

func (self *Node) printNodePointers() {
	node := self.entry.value
	var p, rs, ls, gp, r, l string
	if self.parent == nil {
		p = "no pointer"
		rs = "none"
		ls = "none"
		gp = "none"
	} else {
		if self.parent.right == nil {
			rs = "none"
		} else {
			rs = self.parent.right.entry.value
		}
		if self.parent.left == nil {
			ls = "none"
		} else {
			ls = self.parent.left.entry.value
		}
		if self.parent.parent == nil {
			gp = "none"
		} else {
			gp = self.parent.parent.entry.value
		}
		p = self.parent.entry.value
	}
	if self.right == nil {
		r = "no pointer"
	} else {
		r = self.right.entry.value
	}
	if self.left == nil {
		l = "no pointer"
	} else {
		l = self.left.entry.value
	}
	sibs := (". Right sib: " + rs + ". Left Sib: " + ls + ". grandparent: " + gp)
	essential := ("This is node: " + node + ". Parent: " + p + ". Right: " + r + ". Left: " + l)
	// name := ("I am node: " + self.entry.value + ". ")
	fmt.Println(essential + sibs)
}

func (nd *Node) teachKids() {
	// check if actually need those inputs
	nd.left.parent = nd
	nd.right.parent = nd
}

func (nd *Node) teachParentMyName(is_right bool) {
	if nd.parent != nil {
		if is_right == true {
			nd.parent.right = nd
		} else {
			nd.parent.left = nd
		}
	}
}

type Entry struct {
	key   string
	value string
}

func PublicNewEntry(value string) *Entry {
	h := sha512.New()
	h.Write([]byte(value))
	key := hex.EncodeToString(h.Sum(nil))
	e := Entry{key: key, value: value}
	return &e
}

type MerkleTree struct {
	RootNode  *Node
	RootHash  string
	entries   []*Entry
	entry_map map[*Entry]*Node
	node_map  map[string]*Node
}

func NewMerkleTree() *MerkleTree {
	t := new(MerkleTree)
	t.entry_map = make(map[*Entry]*Node)
	t.node_map = make(map[string]*Node)
	t.RootNode = nil
	t.RootHash = ""
	return t
}

func (t *MerkleTree) makeLeafNode(entry *Entry) *Node {
	t.entries = append(t.entries, entry)
	new_nd := Node{left: nil, right: nil, is_leaf: true}
	new_nd.entry = entry
	t.entry_map[entry] = &new_nd
	t.node_map[entry.key] = &new_nd
	return &new_nd
}

func (t *MerkleTree) updateRoot(nd *Node) string {
	t.RootNode = nd
	t.RootHash = nd.entry.key
	return t.RootHash
}

func (t *MerkleTree) checkIfGrandParentRoot(nd *Node, new_nd *Node) {
	if nd.parent.parent != nil {
		new_nd.parent = nd.parent.parent
	} else {
		t.updateRoot(new_nd)
	}
}

func (t *MerkleTree) Insert(entry *Entry) string {
	new_leaf_nd := t.makeLeafNode(entry)
	if len(t.entry_map) == 1 {
		return t.updateRoot(new_leaf_nd)
	} else if isPowerOfTwo(len(t.entry_map)-1) || len(t.entry_map) == 2 {
		future_sib := t.RootNode
		str := future_sib.entry.key + new_leaf_nd.entry.key
		new_root_hash := getHash(str)
		new_val := future_sib.entry.value + new_leaf_nd.entry.value
		new_root := NewNode(future_sib, new_leaf_nd, false)
		new_root.makeEntry(new_root_hash, new_val)
		// new_root.teachKids(future_sib, new_leaf_nd)
		new_root.teachKids()
		//new_leaf_nd.parent = new_root
		t.node_map[new_root_hash] = &new_root
		fmt.Print("line 186")
		new_root.printNodePointers()
		fmt.Print("line 188")
		new_leaf_nd.printNodePointers()

		return t.updateRoot(&new_root)
	}
	second_to_last_entry := t.entries[len(t.entries)-2]
	temp_root := t.entry_map[second_to_last_entry]
	fmt.Print("line 191")
	temp_root.printNodePointers()

	// initial insert and there is no lonely leaf but not balance
	if temp_root.parent.left.is_leaf == true && temp_root.is_leaf == true {
		str := temp_root.parent.entry.key + new_leaf_nd.entry.key
		new_nd_h := getHash(str)
		new_val := temp_root.parent.entry.value + new_leaf_nd.entry.value
		new_nd := NewNode(temp_root.parent, new_leaf_nd, false)
		new_nd.makeEntry(new_nd_h, new_val)
		t.checkIfGrandParentRoot(temp_root, &new_nd)
		//new_nd.teachKids(temp_root.parent, new_leaf_nd)
		new_nd.teachKids()
		t.node_map[new_nd_h] = &new_nd
		fmt.Print("line 204")
		t.RootNode.printNodePointers()
		temp_root = &new_nd
	} else if temp_root.parent.left.is_leaf == false && temp_root.is_leaf == true {
		str := temp_root.entry.key + new_leaf_nd.entry.key
		new_nd_h := getHash(str)
		new_val := temp_root.entry.value + new_leaf_nd.entry.value
		new_nd := NewNode(temp_root, new_leaf_nd, false)
		new_nd.makeEntry(new_nd_h, new_val)
		new_nd.parent = temp_root.parent
		//new_nd.teachKids(temp_root, new_leaf_nd)
		new_nd.teachKids()
		t.node_map[new_nd_h] = &new_nd
		fmt.Print("line 216")
		t.RootNode.printNodePointers()
		temp_root = &new_nd
	}

	//inserts after the initial insert
	for temp_root != t.RootNode {
		node_to_delete_key := temp_root.parent.entry.key
		str := temp_root.parent.left.entry.key + temp_root.entry.key
		new_nd_h := getHash(str)
		new_val := temp_root.parent.left.entry.value + temp_root.entry.value
		new_nd := NewNode(temp_root.parent.left, temp_root, false)
		new_nd.makeEntry(new_nd_h, new_val)
		t.checkIfGrandParentRoot(temp_root, &new_nd)
		new_nd.teachKids()
		t.node_map[new_nd_h] = &new_nd
		delete(t.node_map, node_to_delete_key)
		temp_root = &new_nd
	}
	// t.RootNode.teachKids()

	return t.updateRoot(temp_root)
}

func (t *MerkleTree) GenerateMerklePath(key string) []string {
	var merkle_path []string
	node, ok := t.node_map[key]
	if !ok {
		err := []string{"path_not_found"}
		return err
	}
	for node != t.RootNode {
		if node.parent.right == node {
			merkle_path = append(merkle_path, node.parent.left.entry.key)
		} else {
			merkle_path = append(merkle_path, node.parent.right.entry.key)
		}
		node = node.parent
	}
	return merkle_path
}

func (t *MerkleTree) VerifyMerklePath(entry *Entry, location int, merkle_path []string) bool {
	return true
}

func (t *MerkleTree) getRightMostNode() *Node {
	right_nd := t.RootNode
	for right_nd.is_leaf == false {
		right_nd = right_nd.right
	}
	return right_nd
}

func (t *MerkleTree) removeRightMostNode(right_nd *Node) *Node {
	if right_nd.parent == t.RootNode {
		t.RootNode = right_nd.parent.left
		//t.RootNode.parent = nil
	} else {
		grandparent := right_nd.parent.parent
		grandparent.right = right_nd.parent.left
		right_nd.parent.left.parent = grandparent
	}
	parent_to_delete := right_nd.parent.entry.key
	delete(t.node_map, parent_to_delete)
	return right_nd
}

func (t *MerkleTree) deleteLeafFromMaps(node *Node) {
	delete(t.entry_map, node.entry)
	delete(t.node_map, node.entry.key)
	removeEntry(t.entries, node.entry)
}

func (t *MerkleTree) Delete(entry *Entry) string {
	delete_nd, ok := t.entry_map[entry]
	fmt.Print("line 299")
	delete_nd.parent.printNodePointers()
	if !ok {
		return "path_not_found"
	}
	right_nd := t.getRightMostNode()
	fmt.Print("line 303")
	right_nd.printNodePointers()
	// Special case 1: only one node in tree prior to delete.
	if len(t.entry_map) == 1 {
		t.RootHash = ""
		t.RootNode = nil
		t.deleteLeafFromMaps(delete_nd)
		return t.RootHash
	} else if len(t.entry_map) == 2 { //Special case 2: 2 leafs in tree prior to delete.
		parent_to_delete := right_nd.parent.entry.key
		if right_nd == delete_nd {
			t.RootNode = right_nd.parent.left
		} else {
			t.RootNode = right_nd
		}
		t.RootNode.left, t.RootNode.right, t.RootNode.parent = nil, nil, nil
		t.RootHash = t.RootNode.entry.key
		t.deleteLeafFromMaps(delete_nd)
		delete(t.node_map, parent_to_delete)
		return t.RootHash
	}

	right_nd = t.removeRightMostNode(right_nd)
	fmt.Print("line 327")
	delete_nd.parent.printNodePointers()

	var temp_node *Node
	if right_nd == delete_nd { // Special case 3: deleting farthest right node and more then 2 entrys
		temp_node = delete_nd.parent.left
	} else {
		temp_node = right_nd
		temp_node.parent = delete_nd.parent
		if delete_nd == delete_nd.parent.right {
			temp_node.parent.right = temp_node
		} else {
			temp_node.parent.left = temp_node
		}
	}
	fmt.Print("line 339")
	temp_node.printNodePointers()
	fmt.Print("line 341")
	delete_nd.parent.printNodePointers()
	for temp_node != t.RootNode {
		var is_right bool
		fmt.Print("line 351")
		temp_node.printNodePointers()
		if temp_node.parent.parent != nil {
			if temp_node.parent.parent == temp_node.parent.parent.right {
				is_right = true
			} else if temp_node.parent == temp_node.parent.parent.left {
				is_right = false
			}
		}
		node_to_delete_key := temp_node.parent.entry.key
		var new_nd Node
		var str string
		var new_val string
		if temp_node == temp_node.parent.right {
			sibling := temp_node.parent.left
			str = sibling.entry.key + temp_node.entry.key
			new_val = sibling.entry.value + temp_node.entry.value
			new_nd = NewNode(temp_node.parent.left, temp_node, false)
		} else if temp_node == temp_node.parent.left {
			sibling := temp_node.parent.right
			str = temp_node.entry.key + sibling.entry.key
			new_val = temp_node.entry.value + sibling.entry.value
			new_nd = NewNode(temp_node, temp_node.parent.right, false)
		}
		new_nd_h := getHash(str)
		new_nd.makeEntry(new_nd_h, new_val)
		t.checkIfGrandParentRoot(temp_node, &new_nd)
		new_nd.teachKids()
		if temp_node.parent != t.RootNode {
			new_nd.teachParentMyName(is_right)
		} else {
			t.RootNode = &new_nd
		}
		t.node_map[new_nd_h] = &new_nd
		delete(t.node_map, node_to_delete_key) //del old parent
		temp_node = &new_nd                    // if new_nd is root level then stop
	}
	return t.RootHash
}

// Below is testing
var COUNT int = 10

func print2DUtil(root *Node, space int) {
	if root == nil {
		return
	}

	space += COUNT
	print2DUtil(root.right, space)

	fmt.Printf("\n")
	for i := COUNT; i < space; i++ {
		fmt.Printf(" ")
	}
	fmt.Printf(root.entry.value)

	print2DUtil(root.left, space)
	println()
}

func print2D(root *Node) {
	print2DUtil(root, 0)
}

func insertTest1() {
	t := NewMerkleTree()
	a := PublicNewEntry("a")
	b := PublicNewEntry("b")
	c := PublicNewEntry("c")
	d := PublicNewEntry("d")
	e := PublicNewEntry("e")
	f := PublicNewEntry("f")
	g := PublicNewEntry("g")

	fmt.Println("test 1 insert a ________")
	t.Insert(a)
	print2D(t.RootNode)
	fmt.Println("test 2 insert b ________")
	t.Insert(b)
	print2D(t.RootNode)
	fmt.Println("test 3 insert c ________")
	t.Insert(c)
	print2D(t.RootNode)
	fmt.Println("test 4 insert d ________")
	t.Insert(d)
	print2D(t.RootNode)

	fmt.Println("test 5 insert e ________")
	t.Insert(e)
	print2D(t.RootNode)
	fmt.Println("test 6 insert f ________")
	t.Insert(f)
	print2D(t.RootNode)
	fmt.Println("test 7 insert g ________")
	t.Insert(g)
	print2D(t.RootNode)

}

func deleteTest0() {
	t := NewMerkleTree()
	a := PublicNewEntry("a")
	b := PublicNewEntry("b")
	c := PublicNewEntry("c")
	t.Insert(a)
	fmt.Println("insert a test 1____________________________________")
	print2D(t.RootNode)

	t.Insert(b)
	fmt.Println("insert b test 2_____________________________________________")
	print2D(t.RootNode)

	t.Insert(c)
	fmt.Println("insert c test 3_____________________________________________")
	print2D(t.RootNode)

	fmt.Println("delete a test 4_____________________________________________")
	t.Delete(a)
	print2D(t.RootNode)
}

func deleteTest1() {
	t := NewMerkleTree()
	a := PublicNewEntry("a")
	b := PublicNewEntry("b")
	c := PublicNewEntry("c")
	d := PublicNewEntry("d")
	e := PublicNewEntry("e")
	f := PublicNewEntry("f")
	g := PublicNewEntry("g")

	//  fmt.Println("test 1 insert a ________")
	t.Insert(a)
	//  fmt.Println("test 2 insert b ________")
	t.Insert(b)
	print2D(t.RootNode)
	//  fmt.Println("test 3 insert c ________")
	t.Insert(c)
	print2D(t.RootNode)
	//  fmt.Println("test 4 insert d ________")
	t.Insert(d)
	//  print2D(t.RootNode)
	fmt.Println("test 5 insert e ________")
	t.Insert(e)
	//  print2D(t.RootNode)
	fmt.Println("test 6 insert f ________")
	t.Insert(f)
	//  print2D(t.RootNode)
	fmt.Println("test 7 insert g ________")
	t.Insert(g)
	print2D(t.RootNode)

	t.Delete(g)
	fmt.Println("delete g test 8_____________________________________________")
	print2D(t.RootNode)
}
func main() {
	//insertTest1()
	//deleteTest1()
	deleteTest0()
}
